timings/lib/timings.dart

```diff
@@ -0,0 +1,225 @@
+
+  void start();
+
+  void stop();
+}
+
+/// Implementation of a real [SyncTimeTracker].
+///
+/// Use [SyncTimeTracker] factory to get an instance.
+class _SyncTimeTrackerImpl extends Timings implements SyncTimeTracker {
+  /// When this operation started, call [start] to set this.
+  @override
+  DateTime get startTime => _startTime;
+  DateTime _startTime;
+
+  /// When this operation stopped, call [stop] to set this.
+  @override
+  DateTime get stopTime => _stopTime;
+  DateTime _stopTime;
+
+  /// Start tracking this operation, must only be called once, before [stop].
+  @override
+  void start() {
+    assert(_startTime == null && _stopTime == null);
+    _startTime = now();
+  }
+
+  /// Stop tracking this operation, must only be called once, after [start].
+  @override
+  void stop() {
+    assert(_startTime != null && _stopTime == null);
+    _stopTime = now();
+  }
+
+  @override
+  T track<T>(T Function() action) {
+    start();
+    try {
+      return action();
+    } finally {
+      stop();
+    }
+  }
+}
+
+/// Implementation of [SyncTimeTracker] that can handle async actions
+///
+/// Async actions returning [Future] will be tracked as single sync time span
+/// from the beginning of execution till completion of future
+///
+/// Use [SyncTimeTracker.kindaAsync] factory to get an instance.
+class _FutureTimeTrackerImpl extends _SyncTimeTrackerImpl {
+  @override
+  T track<T>(T Function() action) {
+    T result;
+    start();
+    try {
+      result = action();
+    } catch (_) {
+      stop();
+      rethrow;
+    }
+    if (result is Future) {
+      return result.whenComplete(stop) as T;
+    } else {
+      stop();
+      return result;
+    }
+  }
+}
+
+/// No-op implementation of [SyncTimeTracker] that does nothing.
+///
+/// Use [SyncTimeTracker.noOp] factory to get an instance.
+class _NoOpSyncTimeTracker implements SyncTimeTracker {
+  static final sharedInstance = _NoOpSyncTimeTracker();
+
+  @override
+  Duration get duration => throw UnimplementedError();
+
+  @override
+  DateTime get startTime => throw UnimplementedError();
+
+  @override
+  DateTime get stopTime => throw UnimplementedError();
+
+  @override
+  void start() {}
+
+  @override
+  void stop() {}
+
+  @override
+  T track<T>(T Function() action) => action();
+}
+
+abstract class AsyncTimeTracker implements AsyncTimings, TimeTracker {
+  factory AsyncTimeTracker() => _AsyncTimeTrackerImpl();
+
+  factory AsyncTimeTracker.noOp() => _NoOpAsyncTimeTracker.sharedInstance;
+}
+
+/// Implementation of a real [AsyncTimeTracker].
+///
+/// Use [AsyncTimeTracker] factory to get an instance.
+class _AsyncTimeTrackerImpl extends AsyncTimings implements AsyncTimeTracker {
+  final internalFutures = <Future>[];
+
+  T _trackSyncSlice<T>(ZoneDelegate parent, Zone zone, T Function() action) {
+    // Don't track time in nested runs twice
+    if (slices.isEmpty || slices.last.stopTime != null) {
```

Comment:
I think this will need to get a little bit more complicated, we need to figure out exactly how to handle the nested zones.

I think what we really want is to _not_ include the time from the nested zone at all in the parent zone. Or, possibly we could even represent that with a special type of TimeSlice?

The primary reasoning for this is that `build_runner` actually does a lot of lazy builds. Basically, inside your Builder code if you read an asset which is generated, that might actually cause it to be built, and we want that to be tracked separately from your Builder.

I don't know if that would be the common case though, or if we might need to provide multiple options?